src/Datadog.Trace/DiagnosticListeners/AspNetCoreDiagnosticObserver.cs

```diff
@@ -0,0 +1,264 @@
+                    OnHostingHttpRequestInStart(arg);
+                    break;
+
+                case "Microsoft.AspNetCore.Hosting.UnhandledException":
+                    OnHostingUnhandledException(arg);
+                    break;
+
+                case "Microsoft.AspNetCore.Hosting.HttpRequestIn.Stop":
+                    OnHostingHttpRequestInStop(arg);
+                    break;
+
+                case "Microsoft.AspNetCore.Mvc.BeforeAction":
+                    OnMvcBeforeAction(arg);
+                    break;
+
+                case "Microsoft.AspNetCore.Mvc.AfterOnActionExecuted":
+                    OnMvcAfterOnActionExecuted(arg);
+                    break;
+            }
+        }
+
+        private static string GetUrl(HttpRequest request)
+        {
+            if (request.Host.HasValue)
+            {
+                return $"{request.Scheme}://{request.Host.Value}{request.PathBase.Value}{request.Path.Value}";
+            }
+
+            // HTTP 1.0 requests are not required to provide a Host to be valid
+            // Since this is just for display, we can provide a string that is
+            // not an actual Uri with only the fields that are specified.
+            // request.GetDisplayUrl(), used above, will throw an exception
+            // if request.Host is null.
+            return $"{request.Scheme}://{NoHostSpecified}{request.PathBase.Value}{request.Path.Value}";
+        }
+
+        private static SpanContext ExtractPropagatedContext(HttpRequest request)
+        {
+            try
+            {
+                // extract propagation details from http headers
+                var requestHeaders = request.Headers;
+
+                if (requestHeaders != null)
+                {
+                    var headersCollection = new DictionaryHeadersCollection();
+
+                    foreach (var header in requestHeaders)
+                    {
+                        string key = header.Key;
+                        string[] values = header.Value.ToArray();
+
+                        if (key != null && values.Length > 0)
+                        {
+                            headersCollection.Add(key, values);
+                        }
+                    }
+
+                    return SpanContextPropagator.Instance.Extract(headersCollection);
+                }
+            }
+            catch (Exception ex)
+            {
+                if (Log.IsEnabled(LogEventLevel.Error))
+                {
+                    Log.Error(ex, "Error extracting propagated HTTP headers.");
+                }
+            }
+
+            return null;
+        }
+
+        private void OnMvcAfterOnActionExecuted(object arg)
+        {
+            if (AfterOnActionActionExecutedContextFetcher.Fetch(arg) is ActionExecutedContext actionExecutedContext &&
+                actionExecutedContext.Exception != null &&
+                !actionExecutedContext.ExceptionHandled)
```

Comment:
What happens with an exception that has not been handled?

Reply:
By default, the web framework handles unhandled exception in user code and returns the configured error page with a 500 status.